tensorzero-internal/src/inference/providers/openai.rs

```diff
@@ -1044,33 +1065,185 @@ pub(super) fn tensorzero_to_openai_system_message<'a>(
-                        content: vec![OpenAIUserContent::Text {
-                            text: Cow::Borrowed(text),
-                        }],
-                    }));
-                }
-                Role::Assistant => {
-                    messages.push(OpenAIRequestMessage::Assistant(
-                        OpenAIAssistantRequestMessage {
-                            content: Some(Cow::Borrowed(text)),
-                            tool_calls: None,
-                        },
-                    ));
-                }
-            },
+            ContentBlock::Text(Text { text }) => {
+                user_content_blocks.push(OpenAIUserContent::Text {
+                    text: Cow::Borrowed(text),
+                });
+            }
+            ContentBlock::ToolCall(_) => {
+                return Err(Error::new(ErrorDetails::InvalidMessage {
+                    message: "Tool calls are not supported in user messages".to_string(),
+                }));
+            }
+            ContentBlock::ToolResult(tool_result) => {
+                messages.push(OpenAIRequestMessage::Tool(OpenAIToolRequestMessage {
+                    content: &tool_result.result,
+                    tool_call_id: &tool_result.id,
+                }));
+            }
+            ContentBlock::Image(ImageWithPath {
+                image,
+                storage_path: _,
+            }) => {
+                user_content_blocks.push(OpenAIUserContent::ImageUrl {
+                    image_url: OpenAIImageUrl {
+                        // This will only produce an error if we pass in a bad
+                        // `Base64Image` (with missing image data)
+                        url: format!("data:{};base64,{}", image.mime_type, image.data()?),
+                    },
+                });
+            }
+            ContentBlock::Thought(_) => {
+                // OpenAI doesn't support thought blocks.
+                // This can only happen if the thought block was generated by another model provider.
+                // At this point, we can either convert the thought blocks to text or drop them.
+                // We chose to drop them, because it's more consistent with the behavior that OpenAI expects.
+
+                // TODO (#1361): warn when we drop thought blocks
+            }
+            ContentBlock::Unknown {
+                data,
+                model_provider_name: _,
+            } => {
+                user_content_blocks.push(OpenAIUserContent::Unknown {
+                    data: Cow::Borrowed(data),
+                });
+            }
+        };
+    }
+
+    // If there are any user content blocks, combine them into a single user message.
+    if !user_content_blocks.is_empty() {
+        messages.push(OpenAIRequestMessage::User(OpenAIUserRequestMessage {
+            content: user_content_blocks,
+        }));
+    }
+
+    Ok(messages)
+}
+
+pub(super) fn tensorzero_to_openai_assistant_messages(
```

Comment:
can we just wrap these in a public method that contains the match statement on the message role?

like in the commented-out code above, there is this:
```
match message.role {
      Role::User => tensorzero_to_openai_user_messages(message),
      Role::Assistant => tensorzero_to_openai_assistant_messages(message),
};
```
We should keep this logic in the implementation and make these functions private so as to not force every implementor of `ProviderTrait` to do this.